src/Orleans.Core/Threading/RecursiveInterlockedExchangeLock.cs

```diff
@@ -0,0 +1,54 @@
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Orleans.Threading
+{
+    /// <summary>
+    /// Lightweight recursive lock.
+    /// </summary>
+    internal sealed class RecursiveInterlockedExchangeLock
+    {
+        private const int Unlocked = -1;
+        private int lockState = Unlocked;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryGet()
+        {
+            var threadId = Thread.CurrentThread.ManagedThreadId;
```

Comment:
Wouldn't it be better to get thread ID once and keep in a `ThreadLocal` variable instead?